src/Microsoft.AspNet.Authentication/ClaimsTransformationOptions.cs

```diff
@@ -3,11 +3,42 @@
 
 using System;
 using System.Security.Claims;
+using System.Threading.Tasks;
 
 namespace Microsoft.AspNet.Authentication
 {
     public class ClaimsTransformationOptions
     {
-        public Func<ClaimsPrincipal, ClaimsPrincipal> Transformation { get; set; }
+        public IClaimsTransformer Transformer { get; set; }
+    }
+
+    public interface IClaimsTransformer
+    {
+        Task<ClaimsPrincipal> TransformAsync(ClaimsPrincipal principal);
+        ClaimsPrincipal Transform(ClaimsPrincipal principal);
+    }
+
+    public class ClaimsTransformer : IClaimsTransformer
+    {
+        public Func<ClaimsPrincipal, Task<ClaimsPrincipal>> TransformAsyncDelegate { get; set; }
```

Comment:
Change this to a ctor that takes only sync, and one that takes both, and no longer make these settable